src/ompl/geometric/planners/aitstar/datastructures/Edge.h

```diff
@@ -0,0 +1,97 @@
+/*********************************************************************
+ * Software License Agreement (BSD License)
+ *
+ *  Copyright (c) 2019-present University of Oxford
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the names of the copyright holders nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ *********************************************************************/
+
+// Authors: Marlin Strub
+
+#ifndef OMPL_GEOMETRIC_PLANNERS_AITSTAR_EDGE_
+#define OMPL_GEOMETRIC_PLANNERS_AITSTAR_EDGE_
+
+#include <array>
+#include <limits>
+#include <memory>
+
+#include "ompl/base/Cost.h"
+
+namespace ompl
+{
+    namespace geometric
+    {
+        namespace aitstar
+        {
+            // Forward declaration of a vertex.
+            class Vertex;
+
+            class Edge
+            {
+            public:
+                /** \brief Needs a default constructor for the binary heap. */
+                Edge();
+
+                /** \brief Constructs an edge from a parent, a child, and the sort key. */
+                Edge(const std::shared_ptr<Vertex> &parent, const std::shared_ptr<Vertex> &child,
+                     const std::array<ompl::base::Cost, 3u> &sortKey);
+
+                /** \brief Destructs an edge. */
+                ~Edge() = default;
+
+                /** \brief Returns the parent in this edge. */
+                std::shared_ptr<Vertex> getParent() const;
```

Comment:
presumably each vertex has only one parent, so maybe change to std::unique_ptr? same comment for every other bit of code that uses shared_ptrs to parents (and children, in cases where there can only be one child).

Reply:
The ownership model I implemented is that the edges share ownership of the vertices with the `ImplicitGraph` (the NN struct in the implicit graph to be specific) and the forward and reverse search trees. An alternative model I thought a lot about was one in which the implicit graph is the sole owner. In this case I think the edges and trees should either hold `std::weak_ptr`s or raw pointers. I think I had `std::weak_ptr`s at some point but changed it to the shared ownership model because the pointers to vertices are indirectly exposed to the public (e.g.  through `AITstar::getEdgesInQueue`) and I don't want the vertices to expire while a user still holds a pointer to them (which could happen when vertices are pruned).
I would suggest keeping the current ownership model of vertices, but if you think a different model is more appropriate I'd be excited to learn about it.

Reply:
Ok, let's leave as-is.